### PR TITLE
Fix #70: Fallback to official upstream if not found in mirrros

### LINKED ISSUE #70: Can not install nightly version with unofficial mirrors
It seems that currently most julia release mirrors do not host nightly builds. As a result, `jill install nightly` errors if I am not using the official mirror.

```
PS D:\code> jill install nightly --upstream SJTUG
JILL - Julia Installer 4 Linux (MacOS, Windows and FreeBSD) -- Light

jill will:
  1) install Julia latest for winnt-x86_64 into C:\Users\melonedo\AppData\Local\julias
  2) make symlinks in C:\Users\melonedo\AppData\Local\julias\bin
You may need to manually add C:\Users\melonedo\AppData\Local\julias\bin to PATH
Continue installation? [Y/n]
----- Download Julia -----
downloading Julia release for latest-winnt-x86_64
failed to find latest-winnt-x86_64 in available upstream. Please try it later.
False
```

### PR DESCRIPTION
**Pull request by melonedo**

### CODE REVIEW

File: jill/download.py [hunk 1]
```diff
@@ -144,11 +144,20 @@ def download_package(version=None, sys=None, arch=None, *,
     print(msg)
     registry = SourceRegistry(upstream=upstream)
     url = registry.query_download_url(version, system, architecture)
-    if not url:
-        msg = f"failed to find {release_str} in available upstream. Please try it later."
+    if not url and upstream != "Official":
```
Comment (crstnbr) [inline jill/download.py:147]: We still want to throw an error message when url is not found and `upstream == Official`, no?

**Review by crstnbr — CHANGES_REQUESTED** [review]

File: jill/download.py (hunk 1 above)
Comment (johnnychen94) [reply jill/download.py:147]: Yes, I believe the fallback workflow is:

1. try the specified upstream, e.g., `SJTUG`
2. if the URL for `SJTUG` doesn't exist, fall back to `Official` with a warning.
3. if the URL for `Official` doesn't exist, exit with an error.

Comment (melonedo) [reply jill/download.py]: Also, the only possible trivial version that can not be found on all upstreams is "nightly" (this is how I tested the error handling), do we need to excluded it in the list of legitimate version strings?

Comment (johnnychen94) [reply jill/download.py]: That might be too delicate, though. I'd prefer to not dealing with these content details in the script.

**Review by johnnychen94 — APPROVED** [review]
preview:

```bash
$ python -m jill download latest --upstream SJTUG
downloading Julia release for latest-mac-x86_64
failed to find latest-mac-x86_64 in upstream SJTUG. Fallback to upstream Official.
downloading from https://julialangnightlies-s3.julialang.org/bin/mac/x64/julia-latest-mac64.dmg
```

```bash
$ python -m jill download 2.0.0
downloading Julia release for 2.0.0-mac-x86_64
failed to find 2.0.0-mac-x86_64 in upstream Official. Please try it later.
```

It's perhaps not very easy to test this, so I'm fine with it no tested. But, if possible, could you add some comments on this workflow? That would be helpful for others to understand it.

File: jill/download.py (hunk 1 above)
Comment (johnnychen94) [reply jill/download.py]: For some extreme situations, if, say, `SJTUG` mirror stops syncing, and Official released 1.7.0. Then we still want to fall back to Official when users call `jill install 1.7.0 --upstream SJTUG`.

File: jill/download.py [hunk 2]
```diff
@@ -142,21 +142,22 @@ def download_package(version=None, sys=None, arch=None, *,
...
         registry = SourceRegistry(upstream=upstream)
         url = registry.query_download_url(version, system, architecture)
         if not url:
-            msg = f"failed to find {release_str} in upstream {upstream}. Please try it later."
+            msg = f"failed to find {release_str} in upstream {upstream}. Fallback to upstream Official."
             logging.warning(msg)
             print(f"{color.RED}{msg}{color.END}")
-            return None
+
+    registry = SourceRegistry(upstream=None if upstream is None else "Official")
```
Comment (crstnbr) [inline jill/download.py:159]: If `upstream != "Official"` and the `query_download_url` call works, i.e. `url != False` it now queries the url twice, which seems unnecessary.

**Review by crstnbr — CHANGES_REQUESTED** [review]

File: jill/download.py (hunk 2 above)
Comment (johnnychen94) [reply jill/download.py:159]: The registry is a singleton object so it only do real-query once; all other queries are directly using the caches.

Comment (johnnychen94) [reply jill/download.py:159]: Oh! I think you're correct here.

More importantly, it always forces the real upstream to either `"Official"` or `None` even if we pass `SJTUG`:

```bash
$ python -m jill download 1.6.0 --upstream BFSU
downloading Julia release for 1.6.0-mac-x86_64
downloading from https://julialang-s3.julialang.org/bin/mac/x64/1.6/julia-1.6.0-mac64.dmg
```

The correct logic should be:

```python
if upstream == "Official or upstream is None:
    # skip step 1 and exit once failed
else:
    # do step 1 and step 2
```

Comment (johnnychen94) [reply jill/download.py:159]: The one I made up is in 6c51ca978020baa3d75f48376272d44226ac374c

**Comment by johnnychen94** [comment]
Given that we've already figured out a solution, I took the liberty to continue this PR in #72, so I'm closing this one in favor of that.

**Pull request closed without merging**